corporate/lib/decorator.py

```diff
@@ -105,6 +105,11 @@ def _wrapped_view_func(
             return HttpResponseRedirect(url)
 
         billing_session = RemoteRealmBillingSession(remote_realm)
+
+        if request.method == "PATCH":
+            # has_request_variables doesn't work with PATCH, so we need to
+            # manually decode the body and set request.POST.
+            request.POST = QueryDict(request.body.decode())
```

Comment:
I don't understand what motivates this change at all; we have `PATCH` in our normal API without AFAIK any hacks like this.

Reply:
I think we do it in here for `rest_path`:
```python
def process_as_post(
    view_func: Callable[Concatenate[HttpRequest, ParamT], HttpResponse]
) -> Callable[Concatenate[HttpRequest, ParamT], HttpResponse]:
    @wraps(view_func)
    def _wrapped_view_func(
        request: HttpRequest, /, *args: ParamT.args, **kwargs: ParamT.kwargs
    ) -> HttpResponse:
        # Adapted from django/http/__init__.py.
        # So by default Django doesn't populate request.POST for anything besides
        # POST requests. We want this dict populated for PATCH/PUT, so we have to
        # do it ourselves.
        #
        # This will not be required in the future, a bug will be filed against
        # Django upstream.

        if not request.POST:
            # Only take action if POST is empty.
            if request.content_type == "multipart/form-data":
                POST, files = MultiPartParser(
                    request.META,
                    BytesIO(request.body),
                    request.upload_handlers,
                    request.encoding,
                ).parse()
                # request.POST is an immutable QueryDict in most cases, while
                # MultiPartParser.parse() returns a mutable instance of QueryDict.
                # This can be fix when https://code.djangoproject.com/ticket/17235
                # is resolved.
                # django-stubs makes QueryDict of different mutabilities incompatible
                # types. There is no way to acknowledge the django-stubs mypy plugin
                # the change of POST's mutability, so we bypass the check with cast.
                # See also: https://github.com/typeddjango/django-stubs/pull/925#issue-1206399444
                POST._mutable = False
                request.POST = cast("_ImmutableQueryDict", POST)
                request.FILES.update(files)
            elif request.content_type == "application/x-www-form-urlencoded":
                request.POST = QueryDict(request.body, encoding=request.encoding)

        return view_func(request, *args, **kwargs)

    return _wrapped_view_func

```

Reply:
Maybe we need to use `process_as_post` decorator here?

Reply:
Seems to be working. I will just push the PR to use it.

Reply:
Yeah that seems a lot happier as a model -- a lot easier to reason about shared code.